server/eclipse-project/src/main/java/us/freeandfair/corla/controller/BallotSelection.java

```diff
@@ -0,0 +1,156 @@
+                         countyID,
+                         BallotManifestInfoQueries::holdingSequenceNumber,
+                         CastVoteRecordQueries::atPosition);
+  }
+
+  /**
+   * same as above with optional dependency injection
+   **/
+  public static List<CVRToAuditResponse>
+      selectBallots(final List<Long> rands,
+                    final Long countyID,
+                    final Function<Long,Optional<BallotManifestInfo>> queryBMI,
+                    final CVRQ queryCVR) {
+
+    // this is what gets added to and returned
+    final List<CVRToAuditResponse> a_list = new LinkedList<CVRToAuditResponse>();
+
+    // dedup! are we supposed to be doing this? probably(?):
+    // ComparisonAuditController does!
+    final List<Long> deduped = dedup(rands);
+
+    Integer audit_sequence_number = 0;
+    for (final Long rand: deduped) {
+      // could we get them all at once? I'm not sure
+      final Optional<BallotManifestInfo> bmiMaybe = queryBMI.apply(rand);
+      CastVoteRecord cvr;
+
+      if (bmiMaybe.isPresent()) {
+        final BallotManifestInfo bmi = bmiMaybe.get();
+        // theoretically we don't need cvrs to render ballot info - practically,
+        // though, we need the cvr info for the app to work
+        cvr = queryCVR.apply(bmi.countyID(),
+                                            bmi.scannerID(),
+                                            bmi.batchID(),
+                                            bmi.ballotPosition(rand));
+        if (cvr == null) {
+          // TODO: create a discrepancy when this happens
+          cvr = notFoundCVR();
+        }
+        a_list.add(toResponse(audit_sequence_number, rand, bmi, cvr));
+      } else {
+        final String msg = "could not find a ballot manifest for random number: "
+            + rand;
+        throw new BallotSelection.MissingBallotManifestException(msg);
+      }
+
+      audit_sequence_number++;
+    }
+    a_list.sort(new BallotOrderComparator());
+    return a_list;
+  }
+
+  /** remove duplicates **/
+  public static List<Long> dedup(final List<Long> rands) {
+    return new LinkedList<Long>(new LinkedHashSet<Long>(rands));
+  }
+
+  /** PHANTOM_RECORD conspiracy theory time **/
+  public static CastVoteRecord notFoundCVR() {
+    final CastVoteRecord cvr = new CastVoteRecord(CastVoteRecord.RecordType.PHANTOM_RECORD,
+                                                  null,
+                                                  0L,
+                                                  0,
+                                                  0,
+                                                  0,
+                                                  "",
+                                                  0,
+                                                  "",
+                                                  "NOT FOUND",
+                                                  null);
+    // TODO prevent the client from requesting info about this cvr
```

Comment:
The client will request cvr/ids/0 and get a 404, which is actually fine but it could be handled better I think. That will probably be worked out in https://www.pivotaltracker.com/story/show/159205773